packages/react-import-remote/src/hooks.ts

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import {useIntersection} from '@shopify/react-intersection-observer';
+import {
+  DeferTiming,
+  WindowWithRequestIdleCallback,
+  RequestIdleCallbackHandle,
+} from '@shopify/async';
+import load from './load';
+
+interface Options<Imported> {
+  nonce?: string;
+  defer?: DeferTiming;
+  onError?(error: Error): void;
+  getImport(window: Window): Imported;
+  onImported?(imported: Imported): void;
+}
+
+type State = 'loading' | 'error' | 'loaded' | null;
+
+export function useImportRemote<Imported = any>(
+  source: string,
+  options: Options<Imported>,
+): {
+  state: State;
+  imported: Imported | null;
+  intersectionRef: any;
+} {
```

Comment:
I am not a big fan of this typing, it technically makes `{state: 'error', imported: Imported}` possible. I'd rather see `{error: Error} | {imported: Imported} | {loading: true}` (you'll probably want to flesh those out so that `error` is `undefined` in all types when no error, `loading` is `false` in all types when not loading, and `imported` is `null` in all types when not loaded)